Feature/openid connect

dylanturn opened the pull request:
Adds OIDC support without needing sidecars/proxies. 

Resolves: https://github.com/cdr/code-server/issues/2636

jawnsy commented on src/node/http.ts:
```diff
@@ -69,6 +71,29 @@ export const authenticated = (req: express.Request): boolean => {
           : req.args.password && safeCompare(req.cookies.key, hash(req.args.password)))
       )
+    case AuthType.Openid:
+      console.log(req.oidc.user)
```
It looks like we might want to use `console.debug` and add some other text here? It's definitely helpful to have debug output as figuring out OIDC integration problems can be... a challenge :sweat_smile: but we'd some other text that we can search for, so we know where the logs are coming from.

maybe something like:

```javascript
console.debug("authenticated using OpenID Connect as", req.oidc.user)
```

other debug info like the returned OIDC claims seems useful too? they're secret credentials, but since code-server is meant for a single user, maybe logging those is OK?

dylanturn replied:
Yes, very sorry, I'll include this change along with the updates to yarn lockfile.

jawnsy replied:
No apologies needed! Thanks so much for your contribution! :heart:

jawnsy reviewed (commented):
Hey there,

Thanks so much for your contribution! This looks really great, but I'm not one of the primary maintainers of code-server, so I'll leave the review to others :smile:

code-asher commented:
Thanks for the contribution! I think we'll need to discuss whether we want to support more authentication methods. The main reason we've been against it is that we felt it wasn't worth the maintenance effort when you could just use a purpose-built proxy.

For example, I personally wish we had recommended an external reverse proxy instead of building one into code-server so I'm reluctant to do a similar thing here.

But I'm definitely curious to everyone's thoughts and this seems to have been requested quite a few times so the desire certainly seems to be there. So maybe we put it to a vote or something.

@jsjoeio @oxy @jawnsy

oxy commented:
Hi there! I'm not really against the PR and I think its a great addition! My only concern is a lack of documentation around how to set up OpenID authentication, what the secrets are, how it would work...

I've never touched OpenID before, and in my quick Googling it didn't sound really easy to me either - so I'd really appreciate if you could also write some documentation on how to set it up :sweat_smile:

dylanturn commented:
> Hi there! I'm not really against the PR and I think its a great addition! My only concern is a lack of documentation around how to set up OpenID authentication, what the secrets are, how it would work...
> 
> I've never touched OpenID before, and in my quick Googling it didn't sound really easy to me either - so I'd really appreciate if you could also write some documentation on how to set it up 😅

Funny you should mention that, I'm busy writing documentation that explains how to setup an OpenID application on Auth0, Okta, and Keycloak. I've also been considering a small code update that would make the `openid-group-claim` argument optional, simplifying the OpenID application configuration process for those who don't require group/role based access.

jsjoeio commented:
First, I just want to say thanks for the contribution! We really appreciate your input and and want to encourage more people to contribute to code-server.

Echoing @code-asher, biggest concern with something like this is:
1. What precedent does it establish for folks who want other types of authentication methods in code-server?
2. What would the maintenance for something like this look like for us a maintainers?

I wish we had some type of plugin ecosystem so that people could build _on top_ of code-server, which would be a nice alternative.

I think we'll need to discuss this as a group before we decide what to do. In the meantime, thank you for being patient!

dylanturn commented:
I'm not really sure what my part in this conversation is meant to be, but for what it's worth here are my thoughts on the two concerns you raised.

> * What precedent does it establish for folks who want other types of authentication methods in code-server?
Would adding LDAP/SAML/Kerberos authentication be problematic provided the implementations were simple and leveraged well known and trusted libraries? 

> * What would the maintenance for something like this look like for us a maintainers?
Given that the implementation uses Auth0s `express-openid-connect` library and only consists of around 100 lines of actual logic, I would imagine this would be maintained like any of the other functionality that leverages 3rd party libraries.

Judging by the contents of `typings/pluginapi.d.ts` and `typings/ipc.d.ts` it looks like there has been work done to create some kind of plugin system. That said, from what I can tell the implemented plugins wouldn't be capable of authenticating users, which makes sense since allowing plugins to authenticate users could prove to be a security risk.

dylanturn commented:
For what it's worth, I just merged in the following changes to this branch:
- Removed the OIDC group claims requirement to allow for simpler OpenID-Connect configurations.
- Added documentation that walks through the process of configuring OpenID-Connect authentication endpoints in Okta, Auth0, and Keycloak.

oxy commented on src/node/http.ts:
```diff
@@ -69,6 +71,39 @@ export const authenticated = (req: express.Request): boolean => {
       )
+    case AuthType.Openid:
+      if (req.oidc.isAuthenticated()) {
+        console.debug("User is authenticated using OpenID Connect\n", req.oidc.user)
```
Thank you so much for your PR! Only nit is that we'd prefer you use `logger.debug` instead of `console.debug` for consistency with the rest of our codebase.

For example: `logger.debug("User authenticated with OpenID Connect", field("user", req.oidc.user))`

Sorry about the `console.debug` suggestion earlier :sweat_smile:

dylanturn replied:
haha, no worries, consistency is king! I'll get it updated and pushed at some point today.

dylanturn replied:
The updates have been made. Let me know if there's anything else that needs to be tweaked.

dylanturn replied:
I pushed one last change to update the logger.debug calls to use field() instead of whatever it is I did.

oxy commented on src/node/http.ts:
```diff
@@ -69,6 +71,39 @@ export const authenticated = (req: express.Request): boolean => {
+          if (key === req.args["openid-group-claim"] && req.args["openid-group-claim"].value) {
+            for (const value in claims) {
+              if (req.args["openid-user-group"] === claims[value]) {
+                console.debug("User is authorized!\n", req.oidc.user)
```
Same as above, `logger.debug`

dylanturn replied:
Done :-)

oxy commented on src/node/http.ts:
```diff
@@ -69,6 +71,39 @@ export const authenticated = (req: express.Request): boolean => {
+        }
+
+        // Throw an error informing the user that they're unauthorized.
+        console.debug("User is not authorized!\n", req.oidc.user)
```
`logger.debug`

dylanturn replied:
Done :-)

oxy commented on src/node/http.ts:
```diff
@@ -69,6 +71,39 @@ export const authenticated = (req: express.Request): boolean => {
+
+      // Returning false means the user isn't authenticated.
+      // This should trigger a redirect so the user can get authenticated.
+      console.debug("User is not authenticated using OpenID Connect\n", req.oidc.user)
```
`logger.debug`

dylanturn replied:
Done :-)

oxy reviewed (changes requested):
Thanks a lot for your contribution! Only nit is the `logger.debug`/`console.debug` difference; sorry about misleading you earlier :sweat_smile:

We've had this feature requested from us several times; I'll test this myself and merge this soon - I figure that since its a relatively isolated change (~150 lines of code), and a library made by a relatively trusted org (Auth0), it wouldn't cause a large maintenance burden.

dylanturn commented:
> Thanks a lot for your contribution! Only nit is the `logger.debug`/`console.debug` difference; sorry about misleading you earlier 😅
> 
> We've had this feature requested from us several times; I'll test this myself and merge this soon - I figure that since its a relatively isolated change (~150 lines of code), and a library made by a relatively trusted org (Auth0), it wouldn't cause a large maintenance burden.

Is there anything you need from me before this can be merged?

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+The purpose of this guide is to demonstrate the process of configuring an OpenID Application as an authentication and authorization source for Code-Server.
+
+More information reguarding how OpenID Connect works can be found here:
```
```suggestion
More information regarding how OpenID Connect works can be found here:
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+# Configuring OpenID Authentication
+
+The purpose of this guide is to demonstrate the process of configuring an OpenID Application as an authentication and authorization source for Code-Server.
```
```suggestion
The purpose of this guide is to demonstrate the process of configuring an OpenID application as an authentication and authorization source for Code-Server.
```

vapurrmaid replied:
```suggestion
The purpose of this guide is to demonstrate the process of configuring an OpenID Application as an authentication and authorization source for code-server.
```

We refer to code-server in lowercase format, see: https://github.com/cdr/styleguide/blob/master/product-names.md#capitalized-terms

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+More information reguarding how OpenID Connect works can be found here:
+
+- https://openid.net/connect/faq/
+- https://auth0.com/docs/protocols/openid-connect-protocol
+- https://developer.okta.com/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc
```
```suggestion
- <https://openid.net/connect/faq/>
- <https://auth0.com/docs/protocols/openid-connect-protocol>
- <https://developer.okta.com/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc>
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+## Prerequisites
+
+- Either Keycloak has been deployed or an existing Auth0/Okta account exist.
```
```suggestion
- Either Keycloak has been deployed or an existing Auth0/Okta account exists.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+## Prerequisites
+
+- Either Keycloak has been deployed or an existing Auth0/Okta account exist.
+  - Auth0 Signup: https://auth0.com/signup
+  - Okta Signup: https://developer.okta.com/signup/
+  - Getting started with Keycloak: https://www.keycloak.org/docs/latest/getting_started/
```
Either `[]()` format, or add `<>` around the URLs

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+  - Auth0 Signup: https://auth0.com/signup
+  - Okta Signup: https://developer.okta.com/signup/
+  - Getting started with Keycloak: https://www.keycloak.org/docs/latest/getting_started/
+- A value needs generated for `openid-secret`. This value will be used to encrypt the OpenID Connect session token.
```
```suggestion
- A value needs to be generated for `openid-secret`. This value will be used to encrypt the OpenID Connect session token.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+  - Getting started with Keycloak: https://www.keycloak.org/docs/latest/getting_started/
+- A value needs generated for `openid-secret`. This value will be used to encrypt the OpenID Connect session token.
+  - MacOS/Linux: `openssl rand -base64 25`
+  - Windows: https://activedirectoryfaq.com/2017/08/creating-individual-random-passwords/
```
```suggestion
  - Windows: <https://activedirectoryfaq.com/2017/08/creating-individual-random-passwords/>
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+### Creating The Application
+
+1. Navigate to the Application management section of your Auth0 dashboard at `https://manage.auth0.com/dashboard/us/{{auth0_account_name}}/applications`
```
```suggestion
1. Navigate to the application management section of your Auth0 dashboard at `https://manage.auth0.com/dashboard/us/{{auth0_account_name}}/applications`
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+### Creating The Application
+
+1. Navigate to the Application management section of your Auth0 dashboard at `https://manage.auth0.com/dashboard/us/{{auth0_account_name}}/applications`
+2. Click the **_Create Application_** button in the top right of the page.
```
```suggestion
2. Click the **Create Application** button in the top right of the page.
```

See: <https://github.com/cdr/styleguide/blob/master/text-formatting.md#bold>

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+1. Navigate to the Application management section of your Auth0 dashboard at `https://manage.auth0.com/dashboard/us/{{auth0_account_name}}/applications`
+2. Click the **_Create Application_** button in the top right of the page.
+3. Either provide a name for this application or use the default, then select **_Regular Web Application_** as the application type and click the blue **_Create_** button.
```
```suggestion
3. Either provide a name for this application or use the default, then select **Regular Web Application** as the application type and click the blue **Create** button.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+### Gather The Client ID
+
+1. Make note of the `Client ID` value. This value will be used in the Code-Server `openid-client-id` configuration variable.
```
```suggestion
1. Make note of the `Client ID` value. This value will be used in the code-server `openid-client-id` configuration variable.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+# Configuring OpenID Authentication
```
```suggestion
# Configuring OpenID authentication
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+  - MacOS/Linux: `openssl rand -base64 25`
+  - Windows: https://activedirectoryfaq.com/2017/08/creating-individual-random-passwords/
+
+## Auth0 OpenID Connect Application Setup
```
```suggestion
## Auth0 OpenID Connect application setup
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+## Auth0 OpenID Connect Application Setup
+
+### Creating The Application
```
```suggestion
### Creating the application
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+   <img src="assets/openid-connect/auth0/create-application.png" />
+   </kbd>
+
+### Gather The Client ID
```
```suggestion
### Gather the client ID
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+   <img src="assets/openid-connect/auth0/update-application.png" />
+   </kbd>
+
+### Example Auth0 Code-Server Configuration
```
```suggestion
### Example Auth0 code-server configuration
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+---
+
+**NOTE:** The `openid-issuer-base-url` should be set to the value of the **_Domain_** field of the Okta application.
```
```suggestion
**NOTE:** The `openid-issuer-base-url` should be set to the value of the **Domain** field of the Okta application.
```

vapurrmaid commented on docs/openid-auth.md:
````diff
@@ -0,0 +1,152 @@
+openid-secret: "{prerequisites_openid-secret_value}"
+```
+
+## Okta OpenID Connect Application Setup
````
```suggestion
## Okta OpenID Connect application setup
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+## Okta OpenID Connect Application Setup
+
+### Creating The Application
```
```suggestion
### Creating the application
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+### Creating The Application
+
+1. Navigate to the **_Add Application_** page of your Okta dashboard at `https://{{okta_account_name}}-admin.okta.com/admin/apps/add-app`
```
```suggestion
1. Navigate to the **Add Application** page of your Okta dashboard at `https://{{okta_account_name}}-admin.okta.com/admin/apps/add-app`
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+### Creating The Application
+
+1. Navigate to the **_Add Application_** page of your Okta dashboard at `https://{{okta_account_name}}-admin.okta.com/admin/apps/add-app`
+2. Click the **_Create New App_** button located in the upper right of the page.
```
```suggestion
2. Click the **Create New App** button located in the upper right of the page.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+1. Navigate to the **_Add Application_** page of your Okta dashboard at `https://{{okta_account_name}}-admin.okta.com/admin/apps/add-app`
+2. Click the **_Create New App_** button located in the upper right of the page.
+3. Set the **_Platform_** to **_Web_** and **_Sign on method_** to **_OpenID Connect_**. Then click **_Create_**
```
```suggestion
3. Set the **Platform** to **Web** and **Sign on method** to **OpenID Connect**. Then click **Create**
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+   <img src="assets/openid-connect/okta/create-application.png" />
+   </kbd>
+
+### Update The Application
```
```suggestion
### Update the application
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+### Update The Application
+
+1. Update the **_Application name_** field with the desired name for this application.
```
```suggestion
1. Update the **Application name** field with the desired name for this application.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+### Update The Application
+
+1. Update the **_Application name_** field with the desired name for this application.
+2. Update the **_Login redirect URIs_** and **_Logout redirect URIs (Optional)_** fields so that they point to the correct code-server endpoint then click **_Save_**
```
```suggestion
2. Update the **Login redirect URIs** and **Logout redirect URIs (Optional)** fields so that they point to the correct code-server endpoint then click **Save**
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+1. Update the **_Application name_** field with the desired name for this application.
+2. Update the **_Login redirect URIs_** and **_Logout redirect URIs (Optional)_** fields so that they point to the correct code-server endpoint then click **_Save_**
+   <kbd><img src="assets/openid-connect/okta/update-application.png"/></kbd>
+3. To update the **_Allowed grant types_** start by navigating to the **_General Settings_** section and clicking the **_Edit_** link in the top right corner of the section card.
```
```suggestion
3. To update the **Allowed grant types** start by navigating to the **General Settings** section and clicking the **Edit** link in the top right corner of the section card.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+2. Update the **_Login redirect URIs_** and **_Logout redirect URIs (Optional)_** fields so that they point to the correct code-server endpoint then click **_Save_**
+   <kbd><img src="assets/openid-connect/okta/update-application.png"/></kbd>
+3. To update the **_Allowed grant types_** start by navigating to the **_General Settings_** section and clicking the **_Edit_** link in the top right corner of the section card.
+4. Once in the edit view, locate the **_Allowed grant types_** checkbox list and make sure the check the boxs for **_Implicit (Hybrid)_** and **_Allow ID Token with implicit grant type_** are checked, then scroll to the bottom of the page and click **_Save_**.
```
```suggestion
4. Once in the edit view, locate the **Allowed grant types** checkbox list and make sure the boxes for **Implicit (Hybrid)** and **Allow ID Token with implicit grant type** are checked, then scroll to the bottom of the page and click **Save**.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+3. To update the **_Allowed grant types_** start by navigating to the **_General Settings_** section and clicking the **_Edit_** link in the top right corner of the section card.
+4. Once in the edit view, locate the **_Allowed grant types_** checkbox list and make sure the check the boxs for **_Implicit (Hybrid)_** and **_Allow ID Token with implicit grant type_** are checked, then scroll to the bottom of the page and click **_Save_**.
+   <kbd><img src="assets/openid-connect/okta/update-application-grants.png"/></kbd>
+5. Lastly, ensure a user is assigned to this application by navigating to the **_Assignments_** tab, clicking on **_Assign_**, then selecting **_Assign to People_**.
```
```suggestion
5. Lastly, ensure a user is assigned to this application by navigating to the **Assignments** tab, clicking on **Assign**, then selecting **Assign to People**.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+5. Lastly, ensure a user is assigned to this application by navigating to the **_Assignments_** tab, clicking on **_Assign_**, then selecting **_Assign to People_**.
+   <kbd><img src="assets/openid-connect/okta/update-application-assignments.png"/></kbd>
+
+### Gather The Client ID
```
```suggestion
### Gather the client ID
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+### Gather The Client ID
+
+1. Make note of the `Client ID` value. This value will be used in the Code-Server `openid-client-id` configuration variable.
```
```suggestion
1. Make note of the `Client ID` value. This value will be used in the code-server `openid-client-id` configuration variable.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+   <img src="assets/openid-connect/okta/application-client-id.png" />
+   </kbd>
+
+### Example Code-Server Configuration
```
```suggestion
### Example code-server configuration
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+---
+
+**NOTE:** The `openid-issuer-base-url` should be set to the value of the **_Okta domain_** field of the Okta application.
```
```suggestion
**NOTE:** The `openid-issuer-base-url` should be set to the value of the **Okta domain** field of the Okta application.
```

vapurrmaid commented on docs/openid-auth.md:
````diff
@@ -0,0 +1,152 @@
+openid-secret: "{prerequisites_openid-secret_value}"
+```
+
+## Keycloak OpenID Connect Client Setup
````
```suggestion
## Keycloak OpenID Connect client setup
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+## Keycloak OpenID Connect Client Setup
+
+### Creating The Client
```
```suggestion
### Creating the client
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+   <img src="assets/openid-connect/keycloak/create-client.png" />
+   </kbd>
+
+### Update The Client Name
```
```suggestion
### Update the client name
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+   <img src="assets/openid-connect/keycloak/update-client.png" />
+   </kbd>
+
+### Example Code-Server Configuration
```
```suggestion
### Example code-server configuration
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+### Creating The Client
+
+1. Once logged into the Keycloak instance, navigate to the **_Clients_** page by clicking the **_Clients_** link in the navigation menu on the left side of the page.
```
```suggestion
1. Once logged into the Keycloak instance, navigate to the **Clients** page by clicking the **Clients** link in the navigation menu on the left side of the page.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+### Creating The Client
+
+1. Once logged into the Keycloak instance, navigate to the **_Clients_** page by clicking the **_Clients_** link in the navigation menu on the left side of the page.
+2. Begin the client creation process by clicking the **_Create_** button located in the top right corner of the clint list.
```
```suggestion
2. Begin the client creation process by clicking the **Create** button located in the top right corner of the client list.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+1. Once logged into the Keycloak instance, navigate to the **_Clients_** page by clicking the **_Clients_** link in the navigation menu on the left side of the page.
+2. Begin the client creation process by clicking the **_Create_** button located in the top right corner of the clint list.
+3. Fill out the client creation fields then click **_Save_** - **_Client ID_**: This is the value that will later be populated in `openid-client-id`. This value is entirely up the the user or process creating the client. - **_Client Protocol_**: This value should be set to **_openid-connect_** - **_Root URL_**: This field should be populated with the Code-Server base url.
```
```suggestion
3. Fill out the client creation fields then click **Save** - Client ID**: This is the value that will later be populated in `openid-client-id`. This value is entirely up the the user or process creating the client. - **Client Protocol**: This value should be set to **openid-connect** - **Root URL**: This field should be populated with the code-server base URL.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+### Update The Client Name
+
+1. Once the **_Save_** button in the **_Add Client_** window has been clicked, the client will be created and the page will be redirected to the client settings view. From inside that view proceed to name the newly create client by populating the **_Name_** field.
```
```suggestion
1. Once the **Save** button in the **Add Client** window has been clicked, the client will be created and the page will be redirected to the client settings view. From inside that view proceed to name the newly create client by populating the **Name** field.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+### Update The Client Name
+
+1. Once the **_Save_** button in the **_Add Client_** window has been clicked, the client will be created and the page will be redirected to the client settings view. From inside that view proceed to name the newly create client by populating the **_Name_** field.
+2. Enable implicit flow by changing **_Implicit Flow Enabled_** from **_OFF_** to **_ON_**.
```
```suggestion
2. Enable implicit flow by changing **Implicit Flow Enabled** from **OFF** to **ON**.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+1. Once the **_Save_** button in the **_Add Client_** window has been clicked, the client will be created and the page will be redirected to the client settings view. From inside that view proceed to name the newly create client by populating the **_Name_** field.
+2. Enable implicit flow by changing **_Implicit Flow Enabled_** from **_OFF_** to **_ON_**.
+3. Scroll to the bottom of the page and click **_Save_**.
```
```suggestion
3. Scroll to the bottom of the page and click **Save**.
```

vapurrmaid commented on docs/openid-auth.md:
```diff
@@ -0,0 +1,152 @@
+
+**NOTE:** The `openid-issuer-base-url` will depend on the Keycloak realm. In the example below the realm is called `master`.
+
+**NOTE:** The `openid-client-id` value should be set to the value given to the **_Client ID_** when the client was first created.
```
```suggestion
**NOTE:** The `openid-client-id` value should be set to the value given to the **Client ID** when the client was first created.
```

vapurrmaid reviewed (commented):
Hey, :clap: thanks so much for a rad contribution. I took a look at _just_ the documentation and have various _very minor nitpicks_ so that it aligns with our style guide at Coder.

Check it out here if you'd like to: <https://github.com/cdr/styleguide>

Namely, we are trying to ensure the following styles:

- code-server is always referred to in lowercase
- headings are sentence case
- UI elements use **bold**, but not italics

:clap:

dylanturn commented:
> Hey, 👏 thanks so much for a rad contribution. I took a look at _just_ the documentation and have various _very minor nitpicks_ so that it aligns with our style guide at Coder.
> 
> Check it out here if you'd like to: https://github.com/cdr/styleguide
> 
> Namely, we are trying to ensure the following styles:
> 
> * code-server is always referred to in lowercase
> * headings are sentence case
> * UI elements use **bold**, but not italics
> 
> 👏

Thank you for making the changes easy to commit :-). I've applied your suggestions, let me know if there's anything else that needs to be updated.

Thank you :-)

vapurrmaid commented:
> > Hey, clap thanks so much for a rad contribution. I took a look at _just_ the documentation and have various _very minor nitpicks_ so that it aligns with our style guide at Coder.
> > Check it out here if you'd like to: https://github.com/cdr/styleguide
> > Namely, we are trying to ensure the following styles:
> > 
> > * code-server is always referred to in lowercase
> > * headings are sentence case
> > * UI elements use **bold**, but not italics
> > 
> > clap
> 
> Thank you for making the changes easy to commit :-). I've applied your suggestions, let me know if there's anything else that needs to be updated.
> 
> Thank you :-)

AFAICT, things are good to go - thanks for taking so much care with this contribution!!!

code-asher commented:
Hey @dylanturn thank you for all the work you've done here!

Since my first comment we've been internally continuing the discussion on whether to accept PRs that introduce alternative authentication methods and we ultimately decided we would stick with the existing policy.

Part of the reason was because we want to limit core features to ones that directly impact the development experience of code-server itself (I feel that authentication is more an infrastructure problem) and the other part was that we didn't feel prepared to own and test the feature especially as we'd need to test manually.

I'm sorry we didn't reach this decision sooner before you put all this work into addressing the comments on your PR! :sob: 

But all is not lost. I think it makes sense to adapt this code to a plugin. I think we can get there most of the way with the existing API but we will probably need to make some changes to do it properly.

I'm not sure exactly what the API will need to look like so we're very much open to ideas. The current API is not considered stable so we can change it as much as we like before finally making it public. Once that's completed we can re-open and adapt the code.
